Add QSPI

maciejbocianski opened the pull request:
### Description

Bring new QSPI HAL API to master.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [X] Feature
    [ ] Breaking change


@bulislaw @donatieng @jamesbeyond @0xc0170 @offirko

0xc0170 committed:
QSPI HAL addition

Adding new QSPI HAL header file. This should help to use memory-maped devices
as memories, graphical displays.

The API consist of few functions, most important are read/write/write_command functions.

The command format is:

```
----------------------------------------------
| Instruction | Address | Alt | Dummy | Data |
----------------------------------------------
```

We define only synch API at the moment.

0xc0170 committed:
QSPI: fix address/alt variable sizes (can be skipped)

Fixing by adding NONE values for both

0xc0170 committed:
QSPI: change length to be in/out parameter

This provides a way to return how many bytes have been written/read (as status codes
are returned via func ret value)

0xc0170 committed:
QSPI: improve mode documentation

SPI mode means Clock polarity and phase mode (0 - 3)

0xc0170 committed:
QSPI: initial HAL nrf52840 version

This commit adds QSPI HAL implementation for nrf52840 MCU targets

0xc0170 committed:
QSPI: fix command declaration names

0xc0170 committed:
QSPI: add STM32 implementation

SenRamakri committed:
Modify QSPI HAL API to include an API for command-transfer operations

SenRamakri committed:
Enabling QSPI headers in Nordic HAL implementation and fix for UART STDIO definitions

SenRamakri committed:
QSPI driver implementation

0xc0170 committed:
QSPI HAL: add disabled flag to format phase

If phase is being skipped, set disabled to true, otherwise false.

0xc0170 committed:
QSPI STM32: fix ssel af selection

0xc0170 committed:
QSPI STM32: fix return value in frequency

0xc0170 committed:
QSPI STM32: set default command values to none

0xc0170 committed:
QSPI STM32: remove polling from write/read

This will be part of custom instruction transfer, the flow will be:

1. write data
2. wait for transfer to complete (poll status register from the memory device)

SenRamakri committed:
Review fixes and doxygen changes

SenRamakri committed:
Fix code style issues

SenRamakri committed:
Add support for 1_1_2 and 1_2_2 modes in HAL

SenRamakri committed:
Changing config and return definitions to adhere to HAL defs

SenRamakri committed:
Remove explicit initialize API and coding style fixes

SenRamakri committed:
Fix bracket placements

SenRamakri committed:
Remove changes to Nordic SDK and modify HAL to track qspi init

SenRamakri committed:
Minor optimizations and code style fixes

SenRamakri committed:
Review fixes

0xc0170 committed:
QSPI STM32: add qspi_command_transfer implementation

0xc0170 committed:
QSPI STM32: init returns error if failed to init

0xc0170 committed:
QSPI STM32: add QSPI_x support to pinnames

0xc0170 committed:
QSPI: fix address size for build qspi command

if address is skipped, used size NONE

0xc0170 committed:
QSPI: remove initialize method

Not used anymore, not defined. ctor default initializes the object

0xc0170 committed:
QSPI: remove spaces on empty lines

0xc0170 committed:
QSPI HAL: fix alternative comment

0xc0170 committed:
QSPI: fix arguments for write/read when alt is defined

Should be: instr, alt, address or inst, addr or just addr

0xc0170 committed:
QSPI: fix driver style issues

0xc0170 committed:
QSPI: fix alt size NONE instead 0

0xc0170 committed:
QSPI STM32: fix disabled format phase

0xc0170 committed:
QSPI STM32: fix pin merging

hw name as input

0xc0170 committed:
QSPI STM32: fix command transfer

use write/read from STM32 driver

0xc0170 committed:
QSPI STM32: fix prepare comman - alt/address

0xc0170 committed:
QSPI STM32: fix default fifo and cycle

As example for DISCO F469NI defines them

0xc0170 committed:
QSPI: add flash pins for F469 disco board

0xc0170 committed:
QSPI: add flash pins for nrf52480_dk board

0xc0170 committed:
QSPI: hal doxygen fixes

Small corrections, fixing typos in the documentation

0xc0170 committed:
QSPI: add STM32L4 support

Disco IoT board support for QSPI. As it does not have dual flash support in QSPI,
we need to fix qspi hal implementation.

0xc0170 committed:
QSPI: add address to command transfer

0xc0170 committed:
QSPI: fix memset header file missing

adustm committed:
Change default FlashSize to 64Mbit = 8Mbytes = 0x800000

adustm committed:
Dummy cycles count is not an init parameter, but a command parameter.
It can change depending on the chosen command, not only at start.
This way we avoid to launch init function and break the internal status variables.

adustm committed:
Fix Address.Size and AlternateByes.Size by shifting them

The ST HAL code is waiting for the correctly shifted vlue
(for a direct write into the HW register)

adustm committed:
Fix Instruction with no data command
Adding QSPI_DATA_NONE activates the transfer
of the command inside HAL_QSPI_COMMAND function

adustm committed:
Move _mode from QSPI::configure_format to QSPI::QSPI
QSPI::configure_format shall not call initialize function.
This way, QSPI::configure_format can be called any time,
only to change the format of commands.
It must contain only parameters for the commands, not for the Init function

adustm committed:
Revert "Dummy cycles count is not an init parameter, but a command parameter."

This reverts commit f1ad089660ac59eb81cfe5d76d7af394d471c30d.

adustm committed:
Enable QSPI feature for DISCO_F413ZH platform

adustm committed:
Support maximum flash size : 4Gbytes

adustm committed:
Add MBED_WEAK for pins

adustm committed:
Enable QSPI for DISCO_F746NG

adustm committed:
Add support for QSPI on DISCO_L476VG

0xc0170 committed:
QSPI: add doxygen options

0xc0170 committed:
QSPI: fix doxy hal documentation

maciejbocianski committed:
fix qspi command transfer for NORDIC

Add address unpacking support.
Only used in erase commands

maciejbocianski committed:
STM qspi: temporary fix for qspi_free return value

maciejbocianski committed:
STM: add qspi pin names for DISCO_L475VG_IOT01A

maciejbocianski committed:
NRF5: fix qspi custom command sending

maciejbocianski committed:
NRF5: fix qspi R/W opcodes mapping

maciejbocianski committed:
add hal-qspi test

maciejbocianski committed:
hal-qspi test refactoring

Main idea of introduced changes is to ease adding support for new flash chips
Major changes:
- move implementation of all memory chip specific functions to memory config file (no weak functions)
- add support for 1-2-2 write

maciejbocianski committed:
hal-qspi test: fix QSPI preprocessor guard

maciejbocianski committed:
nrf52x: fix QSPI enable flag

QSPI feature was mistakenly moved form target nrf52840 to nrf52832
while rebasing. This change fixes it

adustm committed:
Implement qspi_free function

adustm committed:
Fix support of max flash size

jeromecoutant committed:
STM32 : add all QSPI pins in available targets

maciejbocianski committed:
hal-qspi_test: remove STM workaround

after feature-qspi branch rebase, everything works fine and workaround is not needed anymore

adustm committed:
Add reset internal state before call to HAL_QspiInit function

maciejbocianski committed:
hal-qspi test: code refactoring

maciejbocianski committed:
hal-qspi test: add F413ZH support

maciejbocianski committed:
target DISCO_F413ZH: add QSPI flash pin names

offirko commented on TESTS/mbed_hal/qspi/flash_configs/MX25R6435F_config.h:
```diff
@@ -0,0 +1,192 @@
+/* mbed Microcontroller Library
```
Is it common practice to use specific commercial manufacturers setups in mbed Tests?

maciejbocianski replied:
I don't think it's common, actually it's first time. But how to figure out what frequency or what dummy cycle count is supported?

0xc0170 replied:
@OPpuolitaival Please review

OPpuolitaival replied:
I don't know how the configuration works but this sounds wrong place to put board specific configuration. We should have some mechanism existing for these configs

maciejbocianski replied:
Actually it's memory chip specific configuration.The configuration covers: max supported frequency, dummy_cycles_count for different write/read modes, mechanisms for enabling/disabling dual/quad/fast modes (which flag in which register to be set).

Here is example of adding QSPI support to new target by vendor, also contain new memory chip description file ([link](https://github.com/ARMmbed/mbed-os/pull/7825/commits/27f5e7a5eee215881a8ac9808abf5a5a0fcea4cf))

Alternatives:
- put config elsewhere (but where?)
- use SFDP as is in `QSPI SFDP Flash - Block Device` [link](https://github.com/ARMmbed/qspif-blockdevice/pull/1), but it's high level mechanism, not sure if it's adequate here
- reduce test to subset of common features supported by all memory chips (not a good idea, HAL tests should cover max of features available in specific API)

offirko replied:
@maciejbocianski - I have no advise on this subject, sorry.
Other than that looks great to me - thanks.

offirko commented on TESTS/mbed_hal/qspi/flash_configs/N25Q128A_config.h:
```diff
@@ -0,0 +1,182 @@
+    return QSPI_STATUS_OK
+
+
+#define QUAD_ENABLE()                                                           \
```
I see currently those are TODOs, is there a plan to test QUAD?

maciejbocianski replied:
Sure dual/quad will be tested, currently we have problems to enable dual/quad mode on targets with N25Q128A memory chip

offirko commented on TESTS/mbed_hal/qspi/qspi_test_utils.h:
```diff
@@ -0,0 +1,146 @@
+    sUnknown
+};
+
+struct QspiCommand {
```
I know struct and class are both legit, but I would prefer to see here a Class

maciejbocianski replied:
updated, now `QspiCommand` is a class with privat `_cmd`

offirko commented on TESTS/mbed_hal/qspi/qspi_test_utils.h:
```diff
@@ -0,0 +1,146 @@
+
+    qspi_command_t * get();
+
+    qspi_command_t cmd;
```
Another reason to make it a class - If a class would have been used, then mbed-os coding style would have required this member to be '_cmd' - later on in the cpp code its hard to tell whether cmd is a member/global/member

maciejbocianski replied:
updated, now `QspiCommand` is a class with privat `_cmd`

offirko commented on drivers/QSPI.h:
```diff
@@ -0,0 +1,222 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2015 ARM Limited
```
typo I think license date should be updated

offirko commented on drivers/QSPI.h:
```diff
@@ -0,0 +1,222 @@
+
+/** A QSPI Driver, used for communicating with QSPI slave devices
+ *
+ * The default format is set to Quad-SPI(4-4-4), and a clock frequency of 1MHz
```
The default format in the ctor is set to (1-1-1) - QSPI_CFG_BUS_SINGLE

offirko commented on drivers/QSPI.h:
```diff
@@ -0,0 +1,222 @@
+     *
+     *  @param inst_width Bus width used by instruction phase(Valid values are 1,2,4)
+     *  @param address_width Bus width used by address phase(Valid values are 1,2,4)
+     *  @param address_size Size in bits used by address phase(Valid values are 8,16,24,32)
```
same here, better specifies valid ENUM values and not the integer representation which are different

0xc0170 replied:
@maciejbocianski Can you remove those integers. It's obvious what type these parameters are

offirko commented on drivers/QSPI.h:
```diff
@@ -0,0 +1,222 @@
+
+    /** Acquire exclusive access to this SPI bus
+     */
+    virtual void lock(void);
```
Why do we want driver lock/unlock to be public? It could potentially be misused  to block other SPI users.

maciejbocianski replied:
Not sure what was the design @0xc0170 @SenRamakri , but public `lock/unlock` could be utilized to synchronize between two QSPI driver instances controlling two parallel flash chips connected to the same data lines

0xc0170 replied:
Good question. I would have assumed this is made `protected` at least.

@SenRamakri use case for having locks `public`?

offirko commented on drivers/QSPI.cpp:
```diff
@@ -0,0 +1,285 @@
+bool QSPI::_initialize() 
+{
+    if (_mode != 0 && _mode != 1)
+        return QSPI_STATUS_INVALID_PARAMETER;
```
QSPI_STATUS_INVALID_PARAMETER value (-2) is returned as boolean.
Possibly a problem can occur if "acquire" is called and you return without changing _initialized to false.
also single line 'if' should also be within {..}

0xc0170 replied:
AStyle to be run on this feature branch additions

offirko commented on drivers/QSPI.h:
```diff
@@ -0,0 +1,222 @@
+
+    /** Configure the data transmission format
+     *
+     *  @param inst_width Bus width used by instruction phase(Valid values are 1,2,4)
```
Better use the ENUM for Valid values.
The integer representation for the valid values is 0,1,2 : 
typedef enum qspi_bus_width {
    QSPI_CFG_BUS_SINGLE,
    QSPI_CFG_BUS_DUAL,
    QSPI_CFG_BUS_QUAD,
} qspi_bus_width_t;

offirko commented on drivers/QSPI.cpp:
```diff
@@ -0,0 +1,285 @@
+    _initialized = false;
+
+    //Go ahead init the device here with the default config
+    _initialize();
```
ctor can not return potential failure in 'initialize' - maybe we should at least add an error log print?

offirko commented:
@maciejbocianski - Some of my remarks appear twice due to a problem I had in my network during the review - I apologize

stevew817 commented on TESTS/mbed_hal/qspi/qspi_test_utils.cpp:
```diff
@@ -0,0 +1,233 @@
+    for (uint32_t j = 0; j < reg_size; j++) {
+        utest_printf("register byte %u data: ", j);
+        for(int i = 0; i < 8; i++) {
+            utest_printf("%s ", ((reg[j] & (1 << i)) & 0xFF) == 0 ? "0" : "1");
```
Doesn't it make more sense to print the register contents MSB first? I.e. do a `(1 << (7-i))`   instead.

maciejbocianski committed:
mbed_hal-qspi test code refactoring after PR 7783 review

- refactoring of status/config register logging code,
- make QspiCommand a class,

bulislaw commented:
Guys we need to accelerate merging this, we have some dependencies. @donatieng @dannybenor

0xc0170 reviewed (changes requested):
Waiting for the update (@offirko comments addressed)

AnotherButler commented:
Does this docs PR for the driver need updating, or is it still accurate?
https://github.com/ARMmbed/mbed-os-5-docs/pull/354

maciejbocianski committed:
QSPI driver: fixes after PR 7783 review

- make QSPI::lock and QSPI::unlock protected
- fix QSPI::_initialize return value
- add assert on QSPI::_initialize failure
- minor comments fixes

maciejbocianski commented:
Update done all comments were addressed. @offirko please review
astyle fixes on QSPI API/driver/tests added

@donatieng @bulislaw @0xc0170 API has changed `QSPI::lock/unlock` now are protected

maciejbocianski committed:
astyle fixes on QSPI API/driver/tests

donatieng reviewed (approved):
Thanks, looks good, let's get this one in!

jamesbeyond commented:
Hi @AnotherButler, 
The ARMmbed/mbed-os-5-docs#354 is about QSPI drivers API, it looks good, I think @SenRamakri may need to review that again. At least some doxygen links is not valid anymore.

The porting guide is [Here](https://github.com/ARMmbed/mbed-os-5-docs/blob/development/docs/reference/contributing/target/QuadSPI.md), we will update that, and we'll submit a new PR on the handbook.

AnotherButler commented:
@jamesbeyond Thanks

@kegilbert Could you please review the linked PR in place of @SenRamakri as he's out?

maciejbocianski commented:
@bulislaw @AnotherButler @jamesbeyond @donatieng 
Also QSPI porting guide has been updated
https://github.com/ARMmbed/mbed-os-5-docs/pull/676

0xc0170 commented:
@maciejbocianski Please rebase (give it 20 minutes we are currently gatekeeping)

I believe this can be moved to the review (finalize reviews and start CI soon).

maciejbocianski commented:
@0xc0170 rebase done

0xc0170 commented:
/morph build

cmonr commented:
CI related.
/morph build

cmonr commented:
/morph build

kegilbert commented:
Covering for Senthil for the docs portion, left a few comments on the Handbook PRs but overall looks pretty good.

cmonr commented:
/morph build

maciejbocianski commented:
Failed while compiling `mbed-os-example-block device K64F IAR`
```
/SPIFBlockDevice.o ./spif-driver/SPIFBlockDevice.cpp
[Warning] SPIFBlockDevice.h@74,56: [Pe611]: overloaded virtual function "BlockDevice::get_erase_size" is only partially overridden in class "SPIFBlockDevice"
[Error] SPIFBlockDevice.cpp@156,52: [Pe020]: identifier "UINT64_MAX" is undefined
[Error] SPIFBlockDevice.cpp@240,17: [Pe020]: identifier "UINT64_MAX" is undefined
[Warning] SPIFBlockDevice.cpp@449,18: [Pe068]: integer conversion resulted in a change of sign
[Error] SPIFBlockDevice.cpp@557,67: [Pe020]: identifier "UINT64_MAX" is undefined
[Error] SPIFBlockDevice.cpp@877,71: [Pe020]: identifier "UINT64_MAX" is undefined
[Error] SPIFBlockDevice.cpp@922,72: [Pe020]: identifier "UINT64_MAX" is undefined
[Error] SPIFBlockDevice.cpp@942,0: [Pe020]: identifier "UINT64_MAX" is undefined
```

0xc0170 commented:
/morph export-build

0xc0170 commented:
Restarting . Error initialize storage for one target, eclipse.

/morph export-build

cmonr commented:
Pausing Test CI until 5.9.6 PR is merged.
Will restart CI shortly after.

cmonr commented:
Network licensing issues.
/morph export-build

Testing gap while release PR is worked on.
/morph test

cmonr commented:
/morph test

0xc0170 commented:
This PR is ready now for integration.

@AnotherButler Can you review API docs please (there are comments above but not touching  the API docs review, so assuming it has not yet been done) ? If there are any edits needed, can you send them as subsequent PR please (this PR is required by few other dependencies so would like to have this in today to continue other work related).
